vortex-io/src/aligned.rs

```diff
@@ -0,0 +1,161 @@
+#![allow(dead_code)]
+use std::ops::{Deref, DerefMut};
+
+use bytes::Bytes;
+
+pub trait PowerOfTwo<const N: usize> {}
+impl<const N: usize> PowerOfTwo<N> for usize where usize: sealed::Sealed<N> {}
+
+mod sealed {
+    pub trait Sealed<const N: usize> {}
+
+    impl Sealed<1> for usize {}
+    impl Sealed<2> for usize {}
+    impl Sealed<4> for usize {}
+    impl Sealed<8> for usize {}
+    impl Sealed<16> for usize {}
+    impl Sealed<32> for usize {}
+    impl Sealed<64> for usize {}
+    impl Sealed<128> for usize {}
+    impl Sealed<256> for usize {}
+    impl Sealed<512> for usize {}
+}
+
+/// A variant of [`BytesMut`][bytes::BytesMut] that freezes into a [`Bytes`] that is guaranteed
+/// to begin at a multiple of a target byte-alignment.
+///
+/// Internally, it accomplishes this by over-allocating by up to the alignment size, padding the
+/// front as necessary. Reads and writes will only be able to access the region after the padding.
+///
+/// It is required for the alignment to be a valid power of 2 <= 512, any other value will be
+/// a compile-time failure.
+pub(crate) struct AlignedBytesMut<const ALIGN: usize> {
+    buf: Vec<u8>,
+    padding: usize,
+    capacity: usize,
+}
+
+impl<const ALIGN: usize> AlignedBytesMut<ALIGN>
+where
+    usize: PowerOfTwo<ALIGN>,
+{
+    /// Allocate a new mutable buffer with capacity to hold at least `capacity` bytes.
+    ///
+    /// The mutable buffer may allocate more than  the requested amount to pad the memory for
+    /// alignment.
+    pub fn with_capacity(capacity: usize) -> Self {
+        // Allocate up to `ALIGN` extra bytes, in case we need to pad the returned pointer.
+        let allocation_size = (capacity + ALIGN - 1).next_multiple_of(ALIGN);
+        let mut buf = Vec::<u8>::with_capacity(allocation_size);
+        let padding = buf.as_ptr().align_offset(ALIGN);
+        unsafe {
+            buf.set_len(padding);
+        }
+
+        Self {
+            buf,
+            padding,
+            capacity,
+        }
+    }
+
+    /// Usable capacity of this buffer.
+    pub fn capacity(&self) -> usize {
+        self.capacity
```

Comment:
should `capacity` be `allocation_size - padding`? e.g., in Vec and other containers that take `capacity`, the only guarantee is that it will allocate at least that much but by default they tend to over allocate

Reply:
We want to return the original capacity that the user requested. The reason why is because some readers (like the compio one) will use the reported capacity to determine how much data to write.

e.g. if you pass an AlignedBytesMut to `read_exact` you want it to say it has capacity 10, not capacity 64. If you say it has capacity 64, it'll fail the read with unexpectedeof